### PR TITLE
Can now get the latest race results

### PR DESCRIPTION
**Pull request by Evanlab02**

### CODE REVIEW

**Review by Pranav-Zodes-44 — COMMENTED** [review]
What's the difference between the race service and the results service? Will much of the logic vary?
Might it not be better to keep it in one controller? Let me know what you think.

File: Backend/Lyzer/Clients/JolpicaClient.cs [hunk 1]
```diff
@@ -46,5 +46,32 @@ public async Task<DriverStandingsDTO> GetCurrentDriverStandings()
 
             return driverStandings;
         }
+
+        public async Task<ResultsDTO> GetLatestRaceResult()
```
Comment (Pranav-Zodes-44) [inline Backend/Lyzer/Clients/JolpicaClient.cs:50]: Let's also maybe change this to be GetRaceResult with a string year as a parameter. Reduces a need to update it later on when we are wanting to get the race results for a particular year

File: Backend/Lyzer/Services/ResultsService.cs [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+        private readonly CacheService _cache;
+
+        public ResultsService(ILogger<ResultsService> logger, JolpicaClient client, CacheService cache) 
+        {
+            _logger = logger;
+            _client = client;
+            _cache = cache;
+        }
+
+        public async Task<ResultsDTO> GetLatestRaceResult() {
```
Comment (Pranav-Zodes-44) [inline Backend/Lyzer/Services/ResultsService.cs:22]: Let's also maybe change this to be GetRaceResult with a string year as a parameter. Reduces a need to update it later on when we are wanting to get the race results for a particular year. (Same comment as the one that is in jolpica client.
This can then be called in wherever, passing through the year

File: Backend/Lyzer/Controllers/ResultsController.cs [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+        private readonly ILogger<ResultsController> _logger;
+        private readonly ResultsService _resultsService;
+
+        public ResultsController(ILogger<ResultsController> logger, ResultsService resultsService)
+        {
+            _logger = logger;
+            _resultsService = resultsService;
+        }
+
+        [HttpGet("", Name = "GetLatestRaceResult")]
```
Comment (Pranav-Zodes-44) [inline Backend/Lyzer/Controllers/ResultsController.cs:21]: This can also have a parameter of year, current can be a valid year.

Comment (Evanlab02) [reply Backend/Lyzer/Controllers/ResultsController.cs:21]: Hmm so far I do not think we should change this.

We are going to implement a completely different method for retrieving data for previous years in the future as that would be in the database, not Jolpica.

Instead of having to implement checks in the logic to see if it should be using cache or db, I would recommend completely separating concerns by using this endpoint for anything current/latest and then have a different endpoint for previous years data that is something more like `/api/v1/results/<year>`

**Comment by Evanlab02** [comment]
@Pranav-Zodes-44 it is a good question with regards to the difference. 
Initially I was just following the pattern that Jolpica has, differentiating Results and Races.

I think perhaps one could move this to the races controller, however then what Jolpica terms as `Races` should maybe go into something like a `Schedules` controller and I do think for now those should be separated.

I think if we would like, we could combine those into a single endpoint but then we need to implement transformation etc which will definitely be useful/beneficial in future.

I would also be concerned about working too much on a controller that Brendon might be working on causing merge conflicts (which are possibly not that bad, but best to be avoided for now while we can?), and for now, with how small the project is, we can come back to it and adjust once we have better ideas and ironed out first issues.

So in conclusion, I am on board to move this to the Races controller, and then what Brendon does goes into a schedule controller and in future we can pivot and adjust.

File: Backend/Lyzer/Clients/JolpicaClient.cs (hunk 1 above)
Comment (Evanlab02) [reply Backend/Lyzer/Clients/JolpicaClient.cs:50]: So in contrast to my point on the controller.
I am happy to add this parameter, just for future proofing in case we did need it.

So I will add that and adjust to just use current or whatever :)

File: Backend/Lyzer/Services/ResultsService.cs (hunk 2 above)
Comment (Evanlab02) [reply Backend/Lyzer/Services/ResultsService.cs:22]: Same here as with the Client comment.

File: Backend/Lyzer/Controllers/ResultsController.cs (hunk 3 above)
Comment (Evanlab02) [reply Backend/Lyzer/Controllers/ResultsController.cs:21]: Perhaps, what I could do is change this to:

`/api/v1/results/current`

File: Backend/Lyzer/Services/ResultsService.cs (hunk 2 above)
Comment (Evanlab02) [reply Backend/Lyzer/Services/ResultsService.cs:22]: Following up to the conversation on the controller,
I could change change it to get `GetCachedRaceResult` with a parameter in case we ever need to cache results for other years/rounds.

But so far, I am concerned about creating a function called `GetRaceResult` that in future we will build upon with database stuff upon a function that already feels like its doing its fair bit with caching.

File: Backend/Lyzer/Clients/JolpicaClient.cs (hunk 1 above)
Comment (Evanlab02) [reply Backend/Lyzer/Clients/JolpicaClient.cs:50]: Please look at [this](https://github.com/LittleClumsy/Lyzer/pull/41#discussion_r1875500581)

File: Backend/Lyzer/Controllers/ResultsController.cs (hunk 3 above)
Comment (Pranav-Zodes-44) [reply Backend/Lyzer/Controllers/ResultsController.cs:21]: Ok, so what I'm seeing is that you are wanting these endpoints to be always calling Jolpica, and caching the results as well. Then for previous years, they will be accessing the DB.

I think that was where my misunderstanding was. I definitely think `/api/v1/results/current` is better in this case, just to make it a little bit more verbose.

**Comment by Pranav-Zodes-44** [comment]
> @Pranav-Zodes-44 it is a good question with regards to the difference. Initially I was just following the pattern that Jolpica has, differentiating Results and Races.
> 
> I think perhaps one could move this to the races controller, however then what Jolpica terms as `Races` should maybe go into something like a `Schedules` controller and I do think for now those should be separated.
> 
> I think if we would like, we could combine those into a single endpoint but then we need to implement transformation etc which will definitely be useful/beneficial in future.
> 
> I would also be concerned about working too much on a controller that Brendon might be working on causing merge conflicts (which are possibly not that bad, but best to be avoided for now while we can?), and for now, with how small the project is, we can come back to it and adjust once we have better ideas and ironed out first issues.
> 
> So in conclusion, I am on board to move this to the Races controller, and then what Brendon does goes into a schedule controller and in future we can pivot and adjust. What do you think?


Let's it keep it as is for now then, happy to relook at it in the future, and possibly have a quick discussion about it on Wednesday to see how we can go about this in the future? :)

File: Backend/Lyzer/Controllers/ResultsController.cs (hunk 3 above)
Comment (Evanlab02) [reply Backend/Lyzer/Controllers/ResultsController.cs:21]: Perfect yeah, let us chat a bit more on Wednesday.

**Review by LittleClumsy — APPROVED** [review]
Looks good to me!

File: Backend/Lyzer/Clients/JolpicaClient.cs [hunk 4]
```diff
@@ -46,5 +46,32 @@ public async Task<DriverStandingsDTO> GetCurrentDriverStandings()
...
+            JsonElement results = root
+                .GetProperty("MRData")
+                .GetProperty("RaceTable")
+                .GetProperty("Races")[0];
+
+            ResultsDTO? latestResults = JsonConvert.DeserializeObject<ResultsDTO>(results.GetRawText());
+
+            if (latestResults == null)
+            {
+                throw new SerializationException("Could not deserialize driver standings.");
```
Comment (Evanlab02) [inline Backend/Lyzer/Clients/JolpicaClient.cs:71]: Change to correct error message

**Review by Pranav-Zodes-44 — APPROVED** [review]
Happy!
